README.md
# wink-mqtt-rs

This is a rust implementation of an MQTT daemon to run on the wink hub (1, not 2). This turns your wink hub into a mqtt "radio" that can control attached devices.

This version also includes support for [home-assistant autodiscovery](https://www.home-assistant.io/docs/mqtt/discovery/) so you don't have to configure your devices by hand.

There is also a [guide for HomeSeer](https://forums.homeseer.com/forum/hs4-products/hs4-plugins/lighting-primary-technology-plug-ins-aa/mcsmqtt-michael-mcsharry-aa/1455752-hacking-the-wink-hub-v1-to-work-with-homeseer-via-mscmqtt) if you prefer that integration.

## Installation

First you need to have root on your wink hub. [This tutorial](https://www.mattcarrier.com/post/hacking-the-winkhub-part-1/) has instructions on how to root your hub. If you, like me, don't want to buy a UART->USB dongle, you can use the UART port on a Raspberry PI (since both UARTs are 3.3V). This worked for me, but I make no guarantees otherwise.

Once you have root on your hub, run the following command on your wink hub:

```bash
curl -L --cacert /etc/ssl/certs/ca-certificates.crt https://github.com/mikekap/wink-mqtt-rs/releases/latest/download/wink-mqtt-rs.tar.gz | tar -C / -zxvf - && /opt/wink-mqtt-rs/setup.sh
```

and follow the prompts! 

You can configure more options by editing the `/opt/wink-mqtt-rs/config` file after installation (it's just the CLI args to the process).

## Options
```bash
wink-mqtt-rs 0.2.1
Mike Kaplinskiy <mike.kaplinskiy@gmail.com>
wink hub v1 mqtt bridge

USAGE:
    wink-mqtt-rs [FLAGS] [OPTIONS]

FLAGS:
    -h, --help       Prints help information
    -v               verbosity level
    -V, --version    Prints version information

OPTIONS:
        --discovery-listen-topic <discovery-listen-topic>
            Topic to listen to in order to (re)broadcast discovery information. Only applies if
            --discovery-prefix is set. [default: homeassistant/status]

    -d <discovery-prefix>
            Prefix (applied independently of --topic-prefix) to broadcast mqtt discovery information
            (see https://www.home-assistant.io/docs/mqtt/discovery/)

        --http-port <http-port>
            If you'd like an http server, this is the port on which to start it [default: 3000]

    -s <mqtt-uri>
            mqtt server to connect to. Should be of the form
            mqtt[s]://[username:password@]host:port/[?connection_options]

    -i <resync-interval>
            how frequently to check if the light changed state (e.g. via Wink or other external
            means) [default: 10000]

    -t <topic-prefix>
            Prefix for the mqtt topic used for device status/control [default: home/wink/]
```

The default setup above will read these options from `/opt/wink-mqtt-rs/config` . You can also see this by running `cargo +nightly run`.

## Logs

If you're having issues, you can find logs at `/var/log/wink-mqtt-rs.log`. Adding `-vvv` to the config mentioned above will increase the verbosity of logs.

### MQTT Messages

If you have a topic prefix of `home/wink/`, and a device id with `1` named `Fan`:
 - You can *receive* messages on `home/wink/1/status` with the contents, in json:
 ```json
 {"On_Off": 0}
 ```
   The keys/values match the attributes that `aprontest` reports.
 - You can *send* messages on `home/wink/1/set` with the same style json blob as above to set values on the attribute.
 - You can *send* messages on `home/wink/1/7/set` with a value to set for a particular attribute. Note that the attribute id here is a integer as reported by `aprontest`. Prefer the above version in code that does not listen to MQTT Discovery.

Messages on the discovery topic follow a format that works with home assistant MQTT discovery. For details, see [converter.rs](https://github.com/mikekap/wink-mqtt-rs/blob/master/src/converter.rs).

### HTTP Server

An HTTP server is started (by default on port 3000) to let you see a quick UI of what your wink sees. Visit `http://192.168.1.123:3000/` in your browser to see it (replacing `192.168.1.123` with however you reach your wink).

In addition, there's an (unstable) REST API to control the wink exposed via this server. The endpoints are:
```
# List of devices, as well as current attribute values
curl http://wink:3000/api/devices

# Set device id 2's attribute id 3 to 255.
curl http://wink:3000/api/devices/2/3 -d '{"value": 255}' -H "Content-Type: application/json"
```

## Known Issues
 - Groups are not exposed.
 - Does not send device details to Home Assistant, even though the data exists. PRs welcome!
 - `mqtts` support is untested.
 - Don't publish status if nothing changed. Easy fix, if necessary.

## Uninstalling
To uninstall, run:
```bash
/opt/wink-mqtt-rs/uninstall.sh
```

## Developing
This is a vanilla Rust project - just use cargo nightly.

### Running Locally
You can run wink-mqtt-rs locally, though obviously it won't control any lights. There's a fake implementation of aprontest for local use
that mostly just pretends whatever you do to it succeeded.

### Running on the Wink
Use `./release/build_release.sh` to build a ARM binary (requires docker). Then you can:
```bash
ssh root@wink /etc/rc.d/init.d/wink-mqtt-rs stop
scp target/armv5te-unknown-linux-musleabi/release/wink-mqtt-rs root@wink:/opt/wink-mqtt-rs/
ssh root@wink /etc/rc.d/init.d/wink-mqtt-rs start
```

## License

See LICENSE.md


CHANGELOG.md
## 0.2.2
 - Include a device stanza in the autodiscovery payload to create devices (not just entities) in HA.

## 0.2.1
 - Added support for a UINT64 attribute type. (Closes #34)
 - Broadcast discovery for all devices, even if one in the middle fails.
 - Omit attributes from devices with unknown types.

## 0.2.0
 - Add HTTP server running on port 3000. You can toggle stuff on it!
 - Crash when the message queue is full, in the hopes that the supervisor restarts us. rumqttc seems to have issues reconnecting sometimes.
 - Broadcast discovery info every time we connect to the mqtt server.

## 0.1.5
 - Fix STRING types in the JSON mqtt set endpoint as well.

## 0.1.4
 - Add support for older version of firmware that doesn't report Gang data.
 - Add support for STRING-based properties & discovery of STRING-based On_Off switches.
 - Add support for UInt16/32-based properties in parsing.
 - Upgrade the mqtt client to attempt fixing reconnection issues (disconnects from the MQTT server would sometimes fail to retry connecting).

## 0.1.3
 - Report errors more readably better when things fail

## 0.1.2
 - Remove silly Dockerfile in bundle
 - Add version to startup log
 - Avoid burning the CPU if the MQTT server is unreachable

## 0.1.1
 - Add resync interval
 - Add some better log messages

## 0.1.0
 - First release, not in github


src/main.rs
#![feature(async_closure)]

#[macro_use]
extern crate lazy_static;

use std::collections::HashMap;
use std::error::Error;
use std::fs;
use std::io::{BufReader, Read};

use crate::config::Config;
use crate::http::HttpServer;
use clap::{crate_version, App, Arg, ArgMatches, ErrorKind};
use rumqttc::MqttOptions;
use simple_error::bail;
use slog::{info, o, trace, Drain};
use slog_scope::GlobalLoggerGuard;
use slog_term;
use std::sync::Arc;
use tokio::{self, time::Duration};
use url::Url;

mod config;
mod controller;
mod converter;
mod http;
mod syncer;
mod utils;

fn init_logger(args: &ArgMatches) -> GlobalLoggerGuard {
    let min_log_level = match args.occurrences_of("verbose") {
        0 => slog::Level::Info,
        1 => slog::Level::Debug,
        2 | _ => slog::Level::Trace,
    };
    let decorator = slog_term::PlainSyncDecorator::new(std::io::stderr());
    let drain = slog_term::FullFormat::new(decorator)
        .build()
        .filter_level(min_log_level)
        .fuse();
    let logger = slog::Logger::root(drain, o!());
    info!(logger, "init_logger"; "min_log_level" => format!("{:?}", min_log_level));

    let scope_guard = slog_scope::set_global_logger(logger);

    slog_stdlog::init().unwrap();

    scope_guard
}

fn init_mqtt_client(a: &ArgMatches) -> Result<Option<MqttOptions>, Box<dyn Error>> {
    let mqtt_uri = match a.value_of("mqtt-uri") {
        Some(v) => v,
        None => return Ok(None),
    };
    trace!(slog_scope::logger(), "parse_uri"; "uri" => mqtt_uri);
    let mqtt_uri = if !mqtt_uri.starts_with("mqtt://") && !mqtt_uri.starts_with("mqtts://") {
        format!("mqtt://{}", mqtt_uri)
    } else {
        mqtt_uri.to_string()
    };

    let parsed = Url::parse(&mqtt_uri)?;

    if !["mqtt", "mqtts", ""].contains(&parsed.scheme()) {
        bail!("Invalid mqtt url: {}", mqtt_uri)
    }

    let host = match parsed.host() {
        Some(host) => host.to_string(),
        None => bail!("No host in mqtt uri: {}", mqtt_uri),
    };

    let port = parsed.port().unwrap_or(1883);

    let hash_query: HashMap<_, _> = parsed.query_pairs().into_owned().collect();

    let client_id = hash_query
        .get("client_id")
        .map(|x| x.as_str())
        .unwrap_or("wink-mqtt-rs");
    if client_id.starts_with(" ") {
        bail!("Invalid client id: {}", client_id)
    }

    let mut options = MqttOptions::new(client_id, host, port);

    if parsed.username() != "" {
        let password = parsed.password().unwrap_or("");
        options.set_credentials(parsed.username(), password);
    }

    if "mqtts" == parsed.scheme() {
        if let Some(cert) = hash_query.get("tls_root_cert") {
            let mut pem = BufReader::new(fs::File::open(cert)?);
            let mut data = Vec::new();
            pem.read_to_end(&mut data)?;
            options.set_ca(data);
            ()
        } else {
            bail!("Missing root cert for mqtts")
        }
    }

    Ok(Some(options))
}

#[tokio::main]
pub async fn main() -> Result<(), Box<dyn Error>> {
    let matches = App::new("wink-mqtt-rs")
        .version(crate_version!())
        .author("Mike Kaplinskiy <mike.kaplinskiy@gmail.com>")
        .about("wink hub v1 mqtt bridge")
        .arg(Arg::new("verbose")
            .short('v')
            .multiple(true)
            .takes_value(false)
            .about("verbosity level"))
        .arg(Arg::new("resync-interval")
            .short('i')
            .required(false)
            .takes_value(true)
            .about("how frequently to check if the light changed state (e.g. via Wink or other external means)")
            .default_value("10000"))
        .arg(Arg::new("mqtt-uri")
            .short('s')
            .required(false)
            .takes_value(true)
            .about("mqtt server to connect to. Should be of the form mqtt[s]://[username:password@]host:port/[?connection_options]"))
        .arg(Arg::new("topic-prefix")
            .short('t')
            .about("Prefix for the mqtt topic used for device status/control")
            .default_value("home/wink/"))
        .arg(Arg::new("discovery-prefix")
            .short('d')
            .takes_value(true)
            .about("Prefix (applied independently of --topic-prefix) to broadcast mqtt discovery information (see https://www.home-assistant.io/docs/mqtt/discovery/)")
            .required(false))
        .arg(Arg::new("discovery-listen-topic")
            .required(false)
            .takes_value(true)
            .long("--discovery-listen-topic")
            .about("Topic to listen to in order to (re)broadcast discovery information. Only applies if --discovery-prefix is set.")
            .default_value("homeassistant/status"))
        .arg(Arg::new("http-port")
            .required(false)
            .takes_value(true)
            .long("--http-port")
            .about("If you'd like an http server, this is the port on which to start it")
            .default_value("3000"))
        .get_matches();

    let resync_interval: u64 = matches
        .value_of_t("resync-interval")
        .unwrap_or_else(|e| e.exit());

    let http_port = matches
        .value_of_t::<u16>("http-port")
        .map(|t| Some(t))
        .unwrap_or_else(|e| {
            if e.kind == ErrorKind::ArgumentNotFound {
                None
            } else {
                e.exit()
            }
        });

    let _guard = init_logger(&matches);

    info!(slog_scope::logger(), "starting"; "version" => crate_version!());

    let options = init_mqtt_client(&matches)?;
    let config = Config::new(
        options,
        matches.value_of("topic-prefix"),
        matches.value_of("discovery-prefix"),
        matches.value_of("discovery-listen-topic"),
        resync_interval,
        http_port,
    );
    #[cfg(target_arch = "arm")]
    let controller = controller::AprontestController::new();
    #[cfg(not(target_arch = "arm"))]
    let controller = controller::FakeController::new();
    let controller = Arc::new(controller);

    let syncer = if config.has_mqtt() {
        Some(syncer::DeviceSyncer::new(&config, controller.clone()))
    } else {
        None
    };
    let _http = if http_port.is_some() {
        Some(HttpServer::new(&config, controller.clone(), syncer))
    } else {
        None
    };

    loop {
        tokio::time::delay_for(Duration::from_secs(0xfffff)).await;
    }
}
